File: data_loader.py
import tensorflow as tf
import numpy as np
from PIL import Image
import pandas as pd


def load_and_preprocess_img(path):
    img = Image.open(path)
    img = np.array(img)
    img = img / 255.0
    return img

def strArrayToNumpyArray(strArray):
    #takes an string in the form "[1, 2, 3]" and returns a numpy array
    strArray = strArray.replace("[", "")
    strArray = strArray.replace("]", "")
    strArray = strArray.replace(" ", "")
    return np.array(strArray.split(","), dtype=np.int32)

def create_dataset(path="data/tokenized_data/", img_path = "data/preprocessed_imgs/", type="train", batch_size=32, vocab=None):

    df = pd.read_csv(path + "im2latex_"+type+"_tokenized.csv", delimiter=";")
    img_paths = df["image_path"].values
    labels = df["formula"].values


    ds = tf.data.Dataset.from_tensor_slices((img_paths, labels))
    ds = ds.map(lambda x, y: (load_and_preprocess_img(img_path + x), strArrayToNumpyArray(y)))

    return ds.shuffle().batch(batch_size)



#Discarded since tokenizing is now done in the data preprocessing step
""" def tokenize_label(label, vocab, max_len): 
    pad = vocab.token_to_id["<pad>"]
    sos = vocab.token_to_id["<sos>"]
    eos = vocab.token_to_id["<eos>"]
    unk = vocab.token_to_id["<unk>"]

    formula = str(label).strip("\n").split()
    tokenize_formula = []
    tokenize_formula.append(sos)
    for token in formula:
        if token in vocab.tok_to_id:
            tokenize_formula.append(vocab.token_to_id[token])
        else:
            tokenize_formula.append(unk)
    tokenize_formula.append(eos)
    for _ in range(max_len - len(formula)):
        tokenize_formula.append(pad)
    
    return np.array(tokenize_formula) """

File: js_model/data_loader.js
const tf = require("@tensorflow/tfjs-node");
const fs = require("fs").promises;
const path = require("path");
const sharp = require("sharp");
const csv = require("csv-parser");

async function loadAndPreprocessImg(filePath) {
  const imgBuffer = await fs.readFile(filePath);
  const img = tf.node.decodeImage(imgBuffer, 1);
  return img.toFloat().div(tf.scalar(255.0));
}

function strArrayToNumpyArray(strArray) {
  // Takes a string in the form "[1, 2, 3]" and returns a numpy array
  const cleanStrArray = strArray.replace(/\[|\]|\s/g, "");
  const numericArray = cleanStrArray.split(",").map(Number);
  return tf.tensor1d(numericArray, "float32");
}

async function createDataset(
  dataPath = "data/tokenized_data/",
  imgPath = "data/preprocessed_imgs/",
  type = "train",
  batchSize = 32,
  vocab = null
) {
  const csvPath = path.join(dataPath, `im2latex_${type}_tokenized.csv`);
  const records = [];

  // Read CSV file
  await new Promise((resolve, reject) => {
    fs.createReadStream(csvPath)
      .pipe(csv({ delimiter: ";" }))
      .on("data", (record) => {
        records.push(record);
      })
      .on("end", () => {
        resolve();
      })
      .on("error", (error) => {
        reject(error);
      });
  });

  const imgPaths = records.map((record) => record.image_path);
  const labels = records.map((record) => record.formula);

  const ds = tf.data
    .array(imgPaths)
    .mapAsync(async (imgPath, index) => {
      const img = await loadAndPreprocessImg(path.join(imgPath, imgPath));
      const label = strArrayToNumpyArray(labels[index]);
      return [img, label];
    })
    .shuffle(32000)
    .batch(batchSize);

  return ds;
}

// Example usage
const dataset = createDataset();
dataset.forEachAsync((batch) => {
  console.log("Batch:", batch);
});
